Services/TMS/TrainingSearch/classes/class.ilTrainingSearchGUI.php

```diff
@@ -0,0 +1,392 @@
+	 *
+	 * @return BookableCourse[]
+	 */
+	protected function createBookableCourseByFilter(array $crs_infos, array $filter) {
+		$ret = array();
+
+		foreach ($crs_infos as $key => $value) {
+			$crs = $value["crs"];
+
+			$start_date = $crs->getCourseStart();
+			$end_date = $crs->getCourseEnd();
+			$title = $crs->getTitle();
+
+			if($start_date === null) {
+				unset($crs_infos[$key]);
+				continue;
+			}
+
+			list($max_member, $booking_start_date, $booking_end_date, $waiting_list, $min_member, $bookings_available) = $this->helper->getBestBkmValues($value["xbkm"], $start_date);
+			list($venue_id, $city, $address) = $this->helper->getVenueInfos($crs->getId());
+			list($type_id,$type,$target_group_ids,$target_group,$goals,$topic_ids,$topics) = $this->helper->getCourseClassificationValues($value["xccl"]);
+			list($provider_id) = $this->helper->getProviderInfos($crs->getId());
+
+			if(!$this->filter->isInBookingPeriod($booking_start_date, $booking_end_date)) {
+				echo "booking";
+				unset($crs_infos[$key]);
+				continue;
+			}
+
+			if(array_key_exists(self::F_DURATION, $filter)
+				&& !$this->filter->courseInFilterPeriod($start_date, $filter[self::F_DURATION]["start"], $filter[self::F_DURATION]["end"])
+			) {
+				unset($crs_infos[$key]);
+				continue;
+			}
+
+			if(array_key_exists(self::F_TARGET_GROUP, $filter)
+				&& !$this->filter->courseHasTargetGroups($target_group_ids, $filter[self::F_TARGET_GROUP])
+			) {
+				unset($crs_infos[$key]);
+				continue;
+			}
+
+			if(array_key_exists(self::F_TOPIC, $filter)
+				&& !$this->filter->courseHasTopics($topic_ids, $filter[self::F_TOPIC])
+			) {
+				unset($crs_infos[$key]);
+				continue;
+			}
+
+			if(array_key_exists(self::F_TYPE, $filter)
+				&& !$this->filter->courseHasType($type_id, $filter[self::F_TYPE])
+			) {
+				unset($crs_infos[$key]);
+				continue;
+			}
+
+			if(array_key_exists(self::F_TITLE, $filter)
+				&& !$this->filter->crsTitleStartsWith($title, $filter[self::F_TITLE])
+			) {
+				unset($crs_infos[$key]);
+				continue;
+			}
+
+			if(array_key_exists(self::F_NOT_MIN_MEMBER, $filter) 
+				&& $this->filter->minMemberReached($crs->getRefId(), $min_member)
+			) {
+				unset($crs_infos[$key]);
+				continue;
+			}
+
+			if(array_key_exists(self::F_CITY, $filter) 
+				&& $venue_id != -1
+				&& !$this->filter->courseHasVenue($venue_id, $filter[self::F_CITY])
+			) {
+				unset($crs_infos[$key]);
+				continue;
+			}
+
+			if(array_key_exists(self::F_PROVIDER, $filter) 
+				&& !$this->filter->courseHasProvider($provider_id, (int)$filter[self::F_PROVIDER])
+			) {
+				unset($crs_infos[$key]);
+				continue;
+			}
+
+			$start_date_str = $this->helper->formatDate($start_date);
+			$end_date_str = $this->helper->formatDate($end_date);
+
+			$ret[] = $this->db->getBookableCourse($title,
+				$type,
+				$start_date_str,
+				$bookings_available,
+				$target_group,
+				$goals,
+				$topics,
+				$end_date_str,
+				$city,
+				$address
+			);
+		}
```

Comment:
Wenn du diese Magic (und damit auch die Abhängigkeit gegen ilBookableFilter) in die TrainingSearchDB verschiebst, können wir das später viel einfacher auf eine effiziente Suche (über DB, zum Beispiel, oder mit Caching) umstellen.